Turn core API into protocol and implement re-frame.core using a singleton

martinklepsch opened the pull request:
Builds on #1

danielcompton commented on src/re_frame/frame.cljc:
```diff
@@ -0,0 +1,112 @@
+    [re-frame.registry         :as reg]
+    [re-frame.std-interceptors :as stdi]))
+
+(defprotocol IFrame
```
Not sure if this is a good idea or not, but I wonder what would happen if we had `IDispatch`, `ISubscribe`, e.t.c.. Would that open things up for testing more, or is it unnecessary?

martinklepsch replied:
I briefly thought about it but it didn't "jump at me" as the right way to do it. Also wanted to keep the change focused and free from smaller design decisions like this.

danielcompton commented:
This is looking positive. I think the thing that Mike and I have always wondered is how this looks/works in a real application. Would you be able to port the todomvc to pass a frame around to see how it looks?

martinklepsch commented:
@danielcompton The todomvc continues to work with these changes as the core API is maintained through the `re-frame.core/the-frame` singleton. 

Mike mentioned that he is eager to see how this passing around would work and that he'd like to hold off changes like this one before that part is worked out...

The change presented here however isn't aiming to address that. Reagent doesn't provide a way to work with React's Context yet and I don't have any good ideas right now (besides childContext) for that part.

That said this change should make it **a lot** easier for everyone to experiment with their ideas around this problem which to me seems like a logical first step.

danielcompton commented:
Ah of course, I see now.

arichiardi commented:
This honestly looks like a good ol' refactoring, with no API changes, which is good. I have not go deep into the code but I'd like to try it in one of my apps and report back.

martinklepsch commented:
@danielcompton @arichiardi @mike-thompson-day8 

I updated this PR against latest master and implemented a few more changes (notably around the subscription cache and default interceptors).

At this point I would appreciate feedback & suggestions to carry this further. Also just running one or two more complex app on this branch would be of great help to find bugs the tests didn't catch. If that happens I would be happy to extend the test suite as well.

kirillsalykin commented on src/re_frame/core.cljc:
```diff
@@ -43,65 +34,49 @@
+      :event-queue event-queue
+      :app-db re-frame.db/app-db
+      :subs-cache subs-cache
+      :default-interceptors [(cofx/inject-cofx registry :db) (fx/do-fx registry)]})))
```
👍
